### PR TITLE
New context API

### PR DESCRIPTION
**Pull request by rossnomann**
- Use typemap for context.
- Create context for each update (dispatch call) and consume it.
- self is now immutable

### CODE REVIEW

File: src/middleware.rs [hunk 1]
```diff
@@ -48,23 +48,42 @@ impl From<MiddlewareResult> for MiddlewareFuture {
...
+impl ApiMiddleware {
+    /// Creates a new middleware
+    pub fn new(api: Api) -> Self {
+        Self { api }
+    }
+}
+
+impl Middleware for ApiMiddleware {
+    fn before(&mut self, mut context: Context, _update: &Update) -> MiddlewareFuture {
+        context.set(self.api.clone());
```
Comment (ark0f) [inline src/middleware.rs]: Clone and set API every update?

File: src/middleware.rs [hunk 2]
```diff
@@ -107,25 +126,44 @@ mod tests {
...
+    fn handle_message(context: Context, _message: &Message) -> HandlerFuture {
+        context.get::<Counter>().inc_calls();
+        context.into()
+    }
+
+    struct CounterMiddleware;
+
+    impl Middleware for CounterMiddleware {
+        fn before(&mut self, mut context: Context, _update: &Update) -> MiddlewareFuture {
+            context.set(Counter::new());
```
Comment (ark0f) [inline src/middleware.rs]: And here?

File: src/dispatcher.rs [hunk 3]
```diff
@@ -86,19 +81,25 @@ enum DispatcherFutureState {
...
         middleware_error_strategy: ErrorStrategy,
         handler_error_strategy: ErrorStrategy,
         update: Update,
-    ) -> DispatcherFuture<C> {
+    ) -> DispatcherFuture {
         DispatcherFuture {
             middlewares,
             handlers,
-            context,
+            context: Some(context),
```
Comment (ark0f) [inline src/dispatcher.rs]: What is `Option` for?

File: src/dispatcher.rs [hunk 4]
```diff
@@ -243,9 +262,9 @@ impl<C> DispatcherFuture<C> {
     }
 }
 
-impl<C> Future for DispatcherFuture<C> {
-    type Item = ();
-    type Error = Error;
+impl Future for DispatcherFuture {
+    type Item = Context;
+    type Error = (Error, Context);
```
Comment (ark0f) [inline src/dispatcher.rs]: How do you synchronize context?

File: src/dispatcher.rs [hunk 5]
```diff
@@ -196,44 +213,51 @@ impl<C> DispatcherFuture<C> {
...
+            None => {
+                if let Some(context) = self.context.take() {
+                    match self.middlewares.get(idx) {
+                        Some(ref mut middleware) => {
+                            self.middleware = Some(middleware.after(context, &self.update));
+                            task::current().notify();
+                            Ok(Async::NotReady)
+                        }
+                        None => Ok(Async::Ready(context)),
+                    }
```
Comment (mexus) [inline src/dispatcher.rs]: Since the `context` should always be available, I'd suggest to panic if it is not.

File: src/dispatcher.rs [hunk 6]
```diff
@@ -324,36 +357,48 @@ mod tests {
...
-        dispatcher.dispatch(update.clone()).wait().unwrap_err();
-        assert_eq!(dispatcher.context.lock().unwrap().get_calls(), 1);
+        let (_err, context) = dispatcher.dispatch(update.clone()).wait().unwrap_err();
+        assert_eq!(context.get::<Counter>().get_calls(), 1);
 
         // Aborted on handler call
-        let mut dispatcher = Dispatcher::new(
-            vec![Box::new(ErrorMiddleware), Box::new(ErrorMiddleware)],
+        let dispatcher = Dispatcher::new(
+            Api::new("token", None::<&str>).unwrap(),
```
Comment (mexus) [inline src/dispatcher.rs]: It might be a good idea to make a `NoProxy` type just for this case.

Comment (rossnomann) [reply src/dispatcher.rs]: Let's do it in tgbot

File: src/dispatcher.rs (hunk 5 above)
Comment (rossnomann) [reply src/dispatcher.rs]: it panics below

File: src/dispatcher.rs (hunk 6 above)
Comment (mexus) [reply src/dispatcher.rs]: :+1:

File: src/dispatcher.rs (hunk 5 above)
Comment (mexus) [reply src/dispatcher.rs]: Sorry, I've totally lost track in braces :)

File: Cargo.toml [hunk 7]
```diff
@@ -14,6 +14,7 @@ travis-ci = { repository = "tg-rs/carapax", branch = "master" }
 codecov = { repository = "tg-rs/carapax" }
 
 [dependencies]
+anymap = "0.12.1"
```
Comment (ark0f) [inline Cargo.toml]: Set minimal version

**Pull request merged**